File: lab6/c/weight.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <stdarg.h>
#include <stdbool.h>
#include <assert.h>
#include <limits.h>
#include <math.h>

#include "general.h"
#include "weight.h"

///////////////////////// Weights
#define _WINF LONG_MAX
#define _WNINF LONG_MIN


weight_t weight_of_int(signed long w) {
  if (w==_WINF || w==_WNINF) error("Weight overflow");
  return (weight_t){w};
}

weight_t weight_inf() {
  return (weight_t){_WINF };
}

weight_t weight_neg_inf() {
  return (weight_t){_WNINF };
}

weight_t weight_zero() {
  return weight_of_int(0);
}

weight_t weight_of_dbl(double w) {
  return weight_of_int(round(w));
}

bool weight_is_inf(weight_t a) {
  return a.__w == _WINF;
}

bool weight_is_neg_inf(weight_t a) {
  return a.__w == _WNINF;
}

bool weight_is_finite(weight_t a) {
  return !weight_is_inf(a) && !weight_is_neg_inf(a);
}

raw_weight_t weight_to_int(weight_t a) {
  assertmsg(weight_is_finite(a),"Weight must be finite");
  return a.__w;
}





weight_t weight_add(weight_t a, weight_t b) {
  if (weight_is_inf(a)) {
    assertmsg(!weight_is_neg_inf(b), "inf + -inf undefined");
    return weight_inf();
  } else if (weight_is_neg_inf(a)) {
    assertmsg(!weight_is_inf(b), "-inf + inf undefined");
    return weight_neg_inf();
  } else if (weight_is_inf(b)) {
    return weight_inf();
  } else if (weight_is_neg_inf(b)) {
    return weight_neg_inf();
  } else {
    signed long res = a.__w + b.__w; // Not using __builtin_add_overflow as ancient compiler on lab machines does not support it.
//     if (__builtin_add_overflow(a.__w,b.__w,&res)) error("Weight overflow");
    return weight_of_int(res);
  }
}

weight_t weight_sub(weight_t a, weight_t b) {
  if (weight_is_inf(a)) {
    assertmsg(!weight_is_inf(b), "inf - inf undefined");
    return weight_inf();
  } else if (weight_is_neg_inf(a)) {
    assertmsg(!weight_is_neg_inf(b), "-inf - -inf undefined");
    return weight_neg_inf();
  } else if (weight_is_inf(b)) {
    return weight_neg_inf();
  } else if (weight_is_neg_inf(b)) {
    return weight_inf();
  } else {
    signed long res = a.__w - b.__w; // Not using __builtin_sub_overflow as ancient compiler on lab machines does not support it.
//     if (__builtin_sub_overflow(a.__w,b.__w,&res)) error("Weight overflow");
    return weight_of_int(res);
  }
}


bool weight_less(weight_t a, weight_t b) {
  return a.__w < b.__w;
}

bool weight_eq(weight_t a, weight_t b) {
  return a.__w == b.__w;
}


void print_weight(FILE *f, weight_t a) {
  if (weight_is_inf(a)) fprintf(f,"inf");
  else if (weight_is_neg_inf(a)) fprintf(f,"-inf");
  else fprintf(f,"%ld",weight_to_int(a));
}

void print_weight_stdout(weight_t a) {
  if (weight_is_inf(a)) printf("inf");
  else if (weight_is_neg_inf(a)) printf("-inf");
  else printf("%ld",weight_to_int(a));
}

File: lab6/python/weight.py
import sys

import general

WEIGHT_MIN = -sys.maxsize
WEIGHT_MAX = sys.maxsize - 1

_WINF = sys.maxsize
_WNINF = -sys.maxsize - 1

#/// Weight and distance type.

#
 # We provide a weight-type, that implements integers with +infinity and -infinity, and provides
 # overflow-checked operations, which fail on underflows or overflows.
 #
 # It is wrapped into a struct, to prevent accidental access to the raw integer.
 #/
class weight_t:
    def __init__(self,w):
        if (w == _WINF or w == _WNINF):
            general.error("Weight overflow")
        self.w = w

    # Check for +infinity
    def weight_is_inf(self):
        return self.w == _WINF

    # Check for -infinity
    def weight_is_neg_inf(self):
        return self.w == _WNINF

    # Check that weight is finite, i.e., not +inf nor -inf
    def weight_is_finite(self):
        return (not self.weight_is_inf() and not self.weight_is_neg_inf())
        
    def weight_to_int(self):
        assert self.weight_is_finite(), "Weight must be finite"
        return self.w
        
    def print_weight(self, f):
        if (self.weight_is_inf()):
            f.write("inf")
        elif (self.weight_is_neg_inf()):
            f.write("-inf")
        else:
            f.write("%d" % (self.weight_to_int()))

# + infinity
class weight_inf(weight_t):
    def __init__(self):
        self.w = _WINF

# -infinity
class weight_neg_inf(weight_t):
    def __init__(self):
        self.w = _WNINF

# Zero weight
def weight_zero():
    return weight_t(0)

##
 # Add weights. Adding to +inf or -inf will n ot change the weight.
 # Adding +inf and -inf is undefined!
 # @pre {a,b} != {+inf,-inf}
 #/
def weight_add(a, b):
    if (a.weight_is_inf()):
        assert (not b.weight_is_neg_inf()), "inf + -inf undefined"
        return weight_inf()
    elif (a.weight_is_neg_inf()):
        assert(not b.weight_is_inf()), "-inf + inf undefined"
        return weight_neg_inf()
    elif (b.weight_is_inf()):
        return weight_inf()
    elif (b.weight_is_neg_inf()):
        return weight_neg_inf()
    else:
        res = a.w + b.w
        return weight_t(res)

##
 # Subtract weights.
 #
 #    a     b     result
 #    ----------------
 #    inf   inf   undef
 #    inf   *     inf
 #    -inf  -inf  unfef
 #    -inf  *     -inf
 #    fin   inf   -inf
 #    fin   -inf  inf
 #
 #/
def weight_sub(a, b):
    if (a.weight_is_inf()):
        assert (not b.weight_is_inf()), "inf - inf undefined"
        return weight_inf()
    elif (a.weight_is_neg_inf()):
        assert (not b.weight_is_neg_inf()), "-inf - -inf undefined"
        return weight_neg_inf()
    elif (b.weight_is_inf()):
        return weight_neg_inf()
    elif (b.weight_is_neg_inf()):
        return weight_inf()
    else:
        res = a.w - b.w
        return weight_t(res)

# Compare <
def weight_less(a, b):
    return a.w < b.w

# Compare ==
def weight_eq(a, b):
    return a.w == b.w
    
